File: solutions/python/all-your-base/1/all_your_base.py
def validate_inputs(in_base, dgts, out_base):
    if in_base<=1:
        raise ValueError("input base must be >= 2")
    if out_base<=1:
        raise ValueError("output base must be >= 2")
    for d in dgts:
        if d<0 or d>=in_base:
            raise ValueError("all digits must satisfy 0 <= d < input base")

def digits_to_int(base, digits):
    num = 0
    base_power = 1
    digits.reverse()
    for d in digits:
        num += d*base_power
        base_power *= base
    return num 

def int_to_digits(num, base):
    digits = []
    while num>0:
        digits.append(num % base)
        num //= base
    if digits==[]:
        digits = [0]
    digits.reverse()
    return digits

def rebase(input_base, digits, output_base):
    validate_inputs(input_base, digits, output_base)
    number = digits_to_int(input_base, digits)
    result = int_to_digits(number, output_base)
    return result
    

File: solutions/javascript/all-your-base/1/all-your-base.js
//
// all your base #13


const to_value = (digits, base) => {
  let val = 0
  digits.forEach(d => {val = d + val*base})
  return val       
}

const new_digits = (number, base) => {
  let result = []
  while(number>0) {
    result.push(number % base)
    number = Math.floor(number/base)
  }
  return result.reverse()
}

export const convert = (digits, inbase, outbase) => {
  if(inbase<2)
    throw new Error('Wrong input base')
  if(outbase<2)
    throw new Error('Wrong output base')
  if(digits.length == 0  || (digits[0]==0 && digits.length>1) || !(digits.every(d => d>=0)) || !digits.every(d => d<inbase))  
    throw new Error('Input has wrong format')
  if(digits[0] == 0)
    return [0]
  return new_digits(to_value(digits, inbase), outbase)  
};
